### ISSUE TITLE
Magic field `generics` should support user-defined types

### ISSUE THREAD

**Issue by TedDriggs** [TedDriggs/darling#31]
PR #30 by @upsuper added `FromTypeParam`, which enables extraction of attributes on type parameters. Ideally, `darling` would allow a consuming crate to have that extraction done automatically into the master struct implementing `FromDeriveInput`, getting back to the promise of one-line parsing.

- [x] Add `FromGenerics` trait
-  ~`impl<T: FromTypeParam> FromGenerics for Vec<T>`~
- [x] Decide if there needs to be an `ast::Generics` struct
- [x] Decide if `ast::Generics` needs to preserve lifetime and const `GenericParam` values, and if they need to be preserved in order.

@upsuper I'm working on this, but I'd love your input on the decisions listed above.

**Comment by upsuper** [comment]
Ideally it would be good if we would have lifetime and const eventually, so we should probably take them into consider when trying to design stuff here.

For the last point, I don't think we need to preserve the order between different type of params. Currently lifetime must come before type param. Whether const should come after type params doesn't seem to have been decided, but I suspect that would end up being that way. Even if it's not, I don't think it would be a big problem for users.

This would allow us to split generics into different `Vec`s, one for each type of params (rather than needing to have an enum like what `syn` currently does). That probably also indicates `ast::Generics` is probably not worth it, or we may need a generic struct like
```rust
struct<L, T, C> Generics {
    lifetimes: Vec<L>,
    types: Vec<T>,
    consts: Vec<T>,
}
```
It's probably a bit troublesome for people only want some types of params but not others.

Maybe we can just put them into three magic fields, `type_params`, `lifetime_defs`, and `const_params`.

After thinking about `FromGenerics` for a bit, I suspect `impl<T: FromTypeParam> FromGenerics for Vec<T>` may not be a good idea. If we add e.g. `FromLifetimeDef` in the future, we cannot `impl<T: FromLifetimeDef> FromGenerics for Vec<T>` due to conflicting implementations, which would be quite unfortunate.

I'm not sure what can be done for this... Probably we can have `derive(FromTypeParam)` automatically generate a concrete `impl FromGenerics for Vec<SomeType>`?

**Comment by upsuper** [comment]
The `Generics` may not be that bad, though... We can implement `From{TypeParam,LifetimeDef,ConstParam}` for `()` and `!` and users can just use these if they don't want those params.

**Comment by TedDriggs** [comment]
I've removed the blanket impl and replaced it with a generated `impl ... for Vec` when `FromTypeParam` is derived.

I'm also adding a new struct: `darling::util::WithOriginal`. This has two fields: `parsed`, and `original`, which allows someone to write:

```rust
#[derive(FromDeriveInput)]
pub struct MyReceiver {
    generics: darling::util::WithOriginal<Vec<MyTypeParam>, syn::Generics>,
}
```

I think that'll make it easy to pull the relevant parts out of the generics without losing the ability to use `split_for_impl` in your own codegen step.

I'm hoping to ship soon - I _don't_ plan to block 0.4.1 on having `derive(FromGenerics)` since I think this is a very useful alternative. I'm also not planning to include `darling::ast::Generics` in this release; that can wait until we decide on how to deal with lifetimes and const params.

**Comment by TedDriggs** [comment]
I guess I could do a newtype wrapper, like:

```rust
pub struct TypeParams<T>(pub Vec<T>);

impl<T: FromTypeParam> FromGenerics for TypeParams<T> {
    // ... implementation omitted
}
```

It'd be mildly annoying for callers to have to then pierce the newtype wrapper, but that's not the end of the world. I could impl `Into` and `AsRef` on it as well to make it easier to use in places expecting a `Vec`.

**Comment by TedDriggs** [comment]
Fixed by #32
